test/micro/org/openjdk/bench/java/security/PKCS12KeyStores.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.bench.java.security;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.math.BigInteger;
+import java.security.*;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateFactory;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.concurrent.TimeUnit;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+/**
+ * Tests various algorithm settings for PKCS12 keystores.
+ */
+@State(Scope.Thread)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 5)
+@Measurement(iterations = 10)
+@Fork(jvmArgsAppend = {"-Xms1024m", "-Xmx1024m", "-Xmn768m", "-XX:+UseParallelGC"}, value = 5)
+public class PKCS12KeyStores {
+
+    static char[] pass = "changeit".toCharArray();
+
+    static Key pk;
+    static Certificate[] certs;
+    static byte[] bw2048;
+    static byte[] bw50000;
+    static byte[] bs50000;
+    static byte[] bs10000;
+    static byte[] bs2048;
+
+    private static byte[] xeh(String in) {
+        return new BigInteger(in, 16).toByteArray();
+    }
+
+    static {
```

Comment:
Move these to 

```
@Setup
public void setup() throws Exception {
  ...
}
```

This would save you a `try-catch` section, and harness would properly fail on setup exception.

Reply:
Newbie on JMH. Is it worth designing the byte[] fields as static? I assume there will be different objects for different forks but only one for different threads?

Reply:
It does not worth it to make fields `static`. It is much better to let JMH manage the setup, that's what `@Setup` methods are for. If you want to share all instance fields across the thread, then use `@State(Scope.Benchmark)`, it is now `Scope.Thread`.